### ISSUE TITLE
Error Code 20100 on Local Server

### ISSUE THREAD

**Issue by 81457** [barronwaffles/dwc_network_server_emulator#402]
Recently i have installed the server on my Windows 8.1 computer to play WFC on lan, using my 2 wiis.

I installed all the required python dependencies, and the software i use for DNS Hosting is "Dual DHCP DNS Server", i configured the program in DNS mode, and to only listen to the wii's and redirecting all their request from nintendowifi.net to my computer lan ip address.

But when i connect to WFC on TrackMania and Water Warfare! (Both NoSSL Patched) i get error code 20100.

Here are the logs:

```
Dual DHCP DNS Server Version 7.30RC Windows Build 7036 Starting...
Starting DNS Service
DNS Logging: Normal
Server Name: User-PC
Domain Name: 81457sLocalWFC.com
Default Host Expiry: 36000 (sec)
Default Forwarding Server: 192.168.1.1
DNS Service Permitted Hosts: 192.168.1.166-192.168.1.166
Detecting Static Interfaces..
Listening On: 192.168.1.161

# Connecting With Water Warfare Wii (WiiWare)

gpcm.gs.nintendowifi.net A resolved Locally to 192.168.1.161
gpsp.gs.nintendowifi.net A resolved Locally to 192.168.1.161
gamestats.gs.nintendowifi.net A resolved Locally to 192.168.1.161
gamestats2.gs.nintendowifi.net A resolved Locally to 192.168.1.161
wtrwarfarewii.available.gs.nintendowifi.net A resolved Locally to 192.168.1.161
wtrwarfarewii.natneg1.gs.nintendowifi.net A resolved Locally to 192.168.1.161
wtrwarfarewii.natneg2.gs.nintendowifi.net A resolved Locally to 192.168.1.161
wtrwarfarewii.natneg3.gs.nintendowifi.net A resolved Locally to 192.168.1.161
wtrwarfarewii.master.gs.nintendowifi.net A resolved Locally to 192.168.1.161
wtrwarfarewii.gamestats.gs.nintendowifi.net A resolved Locally to 192.168.1.161
wtrwarfarewii.gamestats2.gs.nintendowifi.net A resolved Locally to 192.168.1.161
wtrwarfarewii.ms14.gs.nintendowifi.net A resolved Locally to 192.168.1.161
naswii.nintendowifi.net A resolved Locally to 192.168.1.161

# Connecting With TrackMania (Wii)

naswii.nintendowifi.net A resolved Locally to 192.168.1.161
gpcm.gs.nintendowifi.net A resolved Locally to 192.168.1.161
gpsp.gs.nintendowifi.net A resolved Locally to 192.168.1.161
gamestats.gs.nintendowifi.net A resolved Locally to 192.168.1.161
gamestats2.gs.nintendowifi.net A resolved Locally to 192.168.1.161
trkmaniawii.available.gs.nintendowifi.net A resolved Locally to 192.168.1.161
trkmaniawii.natneg1.gs.nintendowifi.net A resolved Locally to 192.168.1.161
trkmaniawii.natneg2.gs.nintendowifi.net A resolved Locally to 192.168.1.161
trkmaniawii.natneg3.gs.nintendowifi.net A resolved Locally to 192.168.1.161
trkmaniawii.master.gs.nintendowifi.net A resolved Locally to 192.168.1.161
trkmaniawii.gamestats.gs.nintendowifi.net A resolved Locally to 192.168.1.161
trkmaniawii.gamestats2.gs.nintendowifi.net A resolved Locally to 192.168.1.161
trkmaniawii.ms12.gs.nintendowifi.net A resolved Locally to 192.168.1.161
naswii.nintendowifi.net A resolved Locally to 192.168.1.161
gpcm.gs.nintendowifi.net A resolved Locally to 192.168.1.161
gpsp.gs.nintendowifi.net A resolved Locally to 192.168.1.161


# QR Server Prints

[2017-05-05 16:13:50 | GameSpyQRServer] Server is now listening on 0.0.0.0:27900...
[2017-05-05 16:25:25 | GameSpyQRServer] [192.168.1.166:57634] Received availability request for 'wtrwarfarewii' from 192.168.1.166:57634...
[2017-05-05 16:30:25 | GameSpyQRServer] [192.168.1.166:52946] Received availability request for 'trkmaniawii' from 192.168.1.166:52946...
```

Am i doing something wrong or it's an issue with the DNS program itself?

Thanks in advance.

**Comment by sepalani** [comment]
I usually get the error code 20100 when the game can't connect to the NAS server. Be sure you configured Apache (or your HTTP daemon) properly and perform the following port redirections https://github.com/polaris-/dwc_network_server_emulator/tree/master/tools/apache-hosts

Alternatively, you can enable the "AlternativeConfig" in the config file: https://github.com/polaris-/dwc_network_server_emulator/blob/master/altwfc.cfg

With this configuration you won't need to perform the port redirection for the NAS server because it uses the (appropriate) port 80. However, servers running on Linux will need admin privileges to run it. Shouldn't be a problem on Windows.

**Comment by Westretroman** [comment]
Matchmaking server crashed

**Comment by 81457** [comment]
Thanks a lot sepalani!

I got the server working, enabling the Alternative Config and editing the NAS Server ip to my computer's lan ip. All incoming conections are successful and no more EC 20100. :D

I'm going to close this issue, because it's now solved.

**Comment by sepalani** [comment]
@81457 Keep in mind that you'll need to redirect ports at some point if games want to communicate with Gamestats (HTTP) or Sake servers, otherwise you'll get similar troubles.

**Comment by 81457** [comment]
@sepalani Yep, i did the "Alternative Config" just for testing and learning how the server works. I will configure properly the server on Apache once i learn more about it. 

By the way, it's because of ports that TrackMania (Wii) now shows EC 98020 (Local Server) instead of 40255? (EC Shown on AltWFC). I really want to play that game (online) again, but without gamestats server it seems impossible for now, at least an FC can be generated.

Should i re-open this issue? also sorry for the late reply.

**Comment by barronwaffles** [comment]
@81457 There's a commit in my test branch that enables support for most features on trackmania - thankfully it relied on the storage/sake sever rather than gamestats2.

**Comment by 81457** [comment]
@barronwaffles I have just tried your testing branch, and the results are the same (EC 40255). But i see something different in the logs now. 

NasServer used to prints this:
```
Unknown path request /SakeFileServer/upload.aspx?gameid=2793&pid=2
"POST /SakeFileServer/upload.aspx?gameid=2793&pid=2 HTTP/1.1" 404 -
```

And with your commit it now print this:
```
Unknown path request /SakeStorageServer/StorageServer.asmx
"POST /SakeStorageServer/StorageServer.asmx HTTP/1.1" 404 -
```

But the server is still returning 404, maybe that's why the game shows 40255 at login?

I tried creating that directory manually in the server location, but it's no difference.

Also thanks for the reply!
